Foxtrot/Driver/Rewriting/EmitAsyncClosure.cs

```diff
@@ -132,21 +248,45 @@ public EmitAsyncClosure(Method from, Rewriter rewriter)
             this.func2Type = new Cache<TypeNode>(() =>
                 HelperMethods.FindType(SystemTypes.SystemAssembly, StandardIds.System, Identifier.For("Func`2")));
 
-            if (from.IsGeneric)
+            // Should distinguish between generic enclosing method and non-generic method in enclosing type.
+            // In both cases generated closure should be generic.
+            Func<Method, TypeNodeList> getGenericTypesFrom = method =>
+            {
+                if (method.IsGeneric)
+                {
+                    return from.TemplateParameters;
+                }
+                
+                if (method.DeclaringType.IsGeneric)
+                {
+                    return method.DeclaringType.TemplateParameters;
+                }
+
+                return null;
+            };
+
+            var enclosingTemplateParameters = getGenericTypesFrom(from);
```

Comment:
What if method is generic, but returned task uses declaring type generic parameter?
Something like this:
```C#
class DeclaringType<T> where T : class {
    public async Task<T> MethodAsync<U>() {
        Contract.Ensures(Contract.Result<T>() != null);
    }
}
```
or scenario with nested type:
```C#
class DeclaringType<T> where T : class {
    private class NestedType<U> where U : class {
        public async Task<T> MethodAsync<V>() {
            Contract.Ensures(Contract.Result<T>() != null);
        }
    }
}
```
Will this also work?